### PR TITLE
Implement code inclusion

### PR DESCRIPTION
**Pull request by CohenArthur**

### CODE REVIEW

**Comment by CohenArthur** [comment]
A few things are still missing from this PR, such as throwing an error if a folder and a file match the requirements to be included, as well as avoiding double/cyclic inclusions.

I'm looking for reviews on what to improve in order to make the code better/easier to understand

**Comment by CohenArthur** [comment]
@Skallwar @n1tram1 Both features I mentionned are now implemented, please let me know what you think of the implementation :)

File: src/instruction/incl.rs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        let mut dir_fmt = format.clone();
+        let mut file_fmt = format.clone();
+
+        dir_fmt.push_str(&format!("{}", DEFAULT_INCL));
+        file_fmt.push_str(".jk");
+
+        (PathBuf::from(dir_fmt), PathBuf::from(file_fmt))
+    }
+
+    fn format_path(&self, base: &Path) -> Result<PathBuf, JkError> {
```
Comment (Skallwar) [inline src/instruction/incl.rs:46]: I'm not sure about the name. If the goal is to find the file to include, why not ``find_include_path`` or something along those lines ?

File: src/instruction/incl.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        };
+
+        interpreter.debug("BASE DIR", &format!("{:#?}", base));
+
+        let old_path = interpreter.path().cloned();
+
+        let (new_path, content) = self.load(base, interpreter)?;
+
+        // Temporarily change the path of the interpreter
+        interpreter.set_path(Some(new_path));
```
Comment (Skallwar) [inline src/instruction/incl.rs:169]: Why is this needed ?

File: src/parser/constructs.rs [hunk 3]
```diff
@@ -64,6 +65,11 @@ impl Construct {
         Ok((input, expr))
     }
 
+    /// Parse as many instructions as possible
+    pub fn many_instructions(input: &str) -> IResult<&str, Vec<Box<dyn Instruction>>> {
```
Comment (Skallwar) [inline src/parser/constructs.rs:69]: This should return a ``ParseResult<...>``, no ?

File: src/parser/constructs.rs [hunk 4]
```diff
@@ -644,6 +650,49 @@ impl Construct {
         // FIXME: Add Type creation and return it
         Ok((input, ""))
     }
+
+    /// Parses a path for code inclusion
+    fn path(input: &str) -> IResult<&str, String> {
```
Comment (Skallwar) [inline src/parser/constructs.rs:655]: This should return a ``ParseResult<...>``, no ?

File: src/instruction/incl.rs (hunk 2 above)
Comment (CohenArthur) [reply src/instruction/incl.rs:169]: This is needed in order for new includes to be made.

The base is computed from the source's path. 
Let's take an example:

|__ main.jk
|__ std/
    |__ lib.jk
    |__ option/
        |__ std.jk


```rust
// main.jk
incl std
```

```rust
// std/lib.jk
incl option
```

```rust
// std/option/lib.jk
type Option(...);
```

Now let's run `jinko main.jk`

The interpreter computes the parent directory of the source, and looks for a file to include. Here, std/lib.jk fits so we include it.
When going through that file, we see `incl option`. So we need to find a file that fits the requirements. In order to do that, we need to search in std/, not in the parent directory of main.jk

So we need to set the new path of the source before executing its code. If not, then the different source files that that source includes will never be found.
Then, when we're done going through that source file, we set the path back to its original source, so we can keep including from here.

File: src/parser/constructs.rs (hunk 3 above)
Comment (CohenArthur) [reply src/parser/constructs.rs:69]: Yes, you're completely right!

File: src/instruction/incl.rs (hunk 2 above)
Comment (Skallwar) [reply src/instruction/incl.rs:169]: Ok but if there is an error line 177, the interpreter will return with an inconsistent state: line 180 will not have been executed

Comment (CohenArthur) [reply src/instruction/incl.rs:169]: > Ok but if there is an error line 177, the interpreter will return with an inconsistent state: line 180 will not have been executed

Yes, we need to take care of this when we implement the `SpaceLocation` struct. For example, it could keep a tiny reference to a path in order to give errors regarding to the source file the line has been found in, not the source file currently in the interpreter

File: src/instruction/incl.rs (hunk 1 above)
Comment (CohenArthur) [reply src/instruction/incl.rs]: Done

File: src/parser/constructs.rs [hunk 5]
```diff
@@ -672,6 +678,49 @@ impl Construct {
...
+        let (input, path) = Construct::path(input)?;
+
+        let (input, rename) = Construct::as_identifier(input)?;
+
+        let (input, _) = Token::maybe_consume_extra(input)?;
+
+        let incl = Incl::new(path, rename);
+
+        Ok((input, incl))
+    }
```
Comment (Skallwar) [inline src/parser/constructs.rs:723]: ``ParseResult`` ?

**Review by Skallwar — APPROVED** [review]
